File: hospital/SergioDeIscar-hospital-java/SergioDeIscar-hospital-java/src/main/java/org/example/controllers/HospitalController.java
package org.example.controllers;

import org.example.enums.TipoPaciente;
import org.example.exceptions.HospitalExceptions;
import org.example.interfaces.HospitalExtension;
import org.example.models.Paciente;

import java.util.List;

public class HospitalController implements HospitalExtension {
    private final HospitalExtension repo;
    public HospitalController(HospitalExtension repo) {
        this.repo = repo;
    }

    @Override
    public Paciente save(Paciente entity) {
        checkDNI(entity.getDni());
        if (repo.estaCompleto()) { throw new HospitalExceptions.HospitalFullException(); }
        return repo.save(entity);
    }

    @Override
    public Paciente findById(String s) {
        checkDNI(s);
        return repo.findById(s);
    }

    @Override
    public Paciente deleteById(String s) {
        checkDNI(s);
        return repo.deleteById(s);
    }

    @Override
    public List<Paciente> findAll() {
        return repo.findAll();
    }

    @Override
    public void saveAll(List<Paciente> entities) {
        entities.forEach(p -> checkDNI(p.getDni()));
        if (repo.numPacientes() + entities.size() > repo.capacidadMaxima()) { throw new HospitalExceptions.HospitalFullException(); }
        repo.saveAll(entities);
    }

    @Override
    public Boolean estaCompleto() {
        return repo.estaCompleto();
    }

    @Override
    public int numPacientes() {
        return repo.numPacientes();
    }

    @Override
    public List<Paciente> pacientesOrdeFechaIngreso() {
        return repo.pacientesOrdeFechaIngreso();
    }

    @Override
    public List<Paciente> pacientesOrdeNombre() {
        return repo.pacientesOrdeNombre();
    }

    @Override
    public List<Paciente> pacientesOrdeDNI() {
        return repo.pacientesOrdeDNI();
    }

    @Override
    public List<Paciente> pacientesPorTipo(TipoPaciente tipo) {
        return repo.pacientesPorTipo(tipo);
    }

    @Override
    public int numPacientePorTipo(TipoPaciente tipo) {
        return repo.numPacientePorTipo(tipo);
    }

    @Override
    public int capacidadMaxima() {
        return repo.capacidadMaxima();
    }

    private void checkDNI(String dni) {
        if (!dni.matches("[0-9]{8}[A-Z]}")) {
            throw new HospitalExceptions.DniNotValidException();
        }
    }
}

File: hospital/SergioDeIscar-hospital-kotlin/SergioDeIscar-hospital-kotlin/src/main/kotlin/controllers/HospitalController.kt
package controllers

import enums.TipoPaciente
import exceptions.*
import interfaces.HospitalExtension
import models.Paciente

class HospitalController(
    private val repository: HospitalExtension // Inyección de dependencias, se le pasa el repositorio
): HospitalExtension {
    override fun estaCompleto(): Boolean {
        return repository.estaCompleto()
    }

    override fun numPacientes(): Int {
        return repository.numPacientes()
    }

    override fun pacientesOrdeFechaIngreso(): List<Paciente> {
        return repository.pacientesOrdeFechaIngreso()
    }

    override fun pacientesOrdeNombre(): List<Paciente> {
        return repository.pacientesOrdeNombre()
    }

    override fun pacientesOrdeDNI(): List<Paciente> {
        return repository.pacientesOrdeDNI()
    }

    override fun pacientesPorTipo(tipo: TipoPaciente): List<Paciente> {
        return repository.pacientesPorTipo(tipo)
    }

    override fun numPacientePorTipo(tipo: TipoPaciente): Int {
        return repository.numPacientePorTipo(tipo)
    }

    override fun save(paciente: Paciente): Paciente? {
        require(!repository.estaCompleto()) { throw HospitalFullException() }
        require(checkDni(paciente.dni)) {throw DniNotValidException() }
        return repository.save(paciente)
    }

    override fun find(dni: String): Paciente? {
        require(checkDni(dni)) {throw DniNotValidException() }
        return repository.find(dni)
    }

    override fun delete(dni: String): Paciente? {
        require(checkDni(dni)) {throw DniNotValidException() }
        return repository.delete(dni)
    }

    override fun getAll(): List<Paciente> {
        return repository.getAll()
    }

    override fun saveAll(pacientes: List<Paciente>) {
        return repository.saveAll(pacientes)
    }

    private fun checkDni(dni: String): Boolean {
        return Regex("^[0-9]{8}[A-Z]$").matches(dni)
    }
}